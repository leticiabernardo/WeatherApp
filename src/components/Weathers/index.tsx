import WeatherToday from '@/components/Weathers/weatherToday';
import WeatherNextDays from '@/components/Weathers/weatherNextDays';
import { Box } from '@chakra-ui/react';

const Weathers = (): JSX.Element => {
  return (
    <Box
      display="flex"
      alignItems="center"
      justifyContent="center"
      flexDirection="column"
      height="80%"
      width="100%"
    >
      <WeatherToday />
      <WeatherNextDays />
    </Box>
  );
};

export default Weathers;
