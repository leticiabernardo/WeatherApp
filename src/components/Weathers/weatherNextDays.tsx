import { Box, Text } from '@chakra-ui/react';
import { getFullTemperature } from '@/helpers/temperature';
import { getWeekday } from '@/helpers/weekday';
import { getWeatherName } from '@/helpers/weather';
import { useAppContext } from '@/context/Context';
import AsyncSvgIcon from '@/components/AsyncSvgIcon';

const WeatherNextDays = () => {
  const { nextWeathers, temperatureMetric } = useAppContext();

  return (
    <Box
      display="grid"
      gridTemplateColumns={{
        base: '1fr 1fr 1fr 1fr 1fr 1fr 1fr',
        sm: '1fr 1fr 1fr 1fr',
        lg: '1fr 1fr 1fr 1fr 1fr 1fr 1fr',
      }}
      color="white"
      width="100%"
      padding="20px 0"
      margin="0"
      justifyContent="center"
      borderTop="1px solid white"
    >
      {nextWeathers.map(weatherDay => {
        return (
          <Box
            margin="5px 0"
            padding="0 25px"
            display="flex"
            flexDirection="column"
            alignItems="center"
            key={weatherDay.dt}
          >
            <Text fontWeight="bold" textShadow="1px 1px 5px rgba(0,0,0,0.3)">
              {weatherDay.dt ? getWeekday(weatherDay.dt * 1000) : ''}
            </Text>

            <Box
              margin="15px 0"
              display="flex"
              flexDirection="column"
              alignItems="center"
            >
              <AsyncSvgIcon
                svg={`weather-${weatherDay.weather.toLocaleLowerCase()}`}
              />
              <Text marginTop="8px" textShadow="1px 1px 5px rgba(0,0,0,0.3)">
                {getWeatherName(weatherDay.weather)}
              </Text>
            </Box>

            <Text
              textShadow="1px 1px 5px rgba(0,0,0,0.3)"
              fontSize="sm"
              textAlign="center"
            >
              <Text>
                {`Max: ${getFullTemperature(
                  weatherDay.temp.max,
                  temperatureMetric
                )}`}
              </Text>
              <Text>
                {`Min: ${getFullTemperature(
                  weatherDay.temp.min,
                  temperatureMetric
                )}`}
              </Text>
            </Text>
          </Box>
        );
      })}
    </Box>
  );
};

export default WeatherNextDays;
