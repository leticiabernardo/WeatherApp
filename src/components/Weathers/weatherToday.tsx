import { useEffect, useState } from 'react';
import { Box, Text } from '@chakra-ui/react';
import { format } from 'date-fns';
import { useAppContext } from '@/context/Context';
import { getWeatherLocation } from '@/helpers/weather';
import { useTranslation } from 'react-i18next';
import { getDate } from '@/helpers/weekday';
import WeatherTodayDetails from './weatherTodayDetails';

const WeatherToday = (): JSX.Element => {
  const { t, i18n } = useTranslation();
  const [location, setLocation] = useState<string | undefined>(undefined);
  const { location: currLocation } = useAppContext();

  useEffect(() => {
    if (currLocation) {
      setLocation(getWeatherLocation(currLocation, 'long'));
    }
  }, [currLocation]);

  return (
    <div>
      {location && (
        <Box width="100%">
          <Box
            width="100%"
            padding="20px 0 0"
            textAlign={{ base: 'left', sm: 'center', lg: 'left' }}
          >
            <Text
              fontSize="7xl"
              color="white"
              fontFamily="Open Sans"
              lineHeight="65px"
              textShadow="1px 1px 5px rgba(0,0,0,0.3)"
            >
              {location}
            </Text>
            <Text
              fontSize="1xl"
              color="white"
              fontWeight="bold"
              fontFamily="Open Sans"
              textShadow="1px 1px 5px rgba(0,0,0,0.3)"
            >
              {format(new Date(), 'HH:ii')}
            </Text>
            <Text
              fontSize="sm"
              color="white"
              fontFamily="Open Sans"
              textShadow="1px 1px 5px rgba(0,0,0,0.3)"
            >
              {t('Today {{date}}', {
                date: getDate(new Date(), i18n.language),
                interpolation: { escapeValue: false },
              })}
            </Text>
          </Box>
          <Box textAlign={{ base: 'left', sm: 'center', lg: 'left' }}>
            <WeatherTodayDetails />
          </Box>
        </Box>
      )}
    </div>
  );
};

export default WeatherToday;
