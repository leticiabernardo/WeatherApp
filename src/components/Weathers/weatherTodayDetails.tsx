import { useTranslation } from 'react-i18next';
import { Box, Text, Skeleton, SkeletonText } from '@chakra-ui/react';
import { getFullTemperature } from '@/helpers/temperature';
import AsyncSvgIcon from '@/components/AsyncSvgIcon';

const WeatherTodayDetails = (props: {
  weather: app.CurrentWeather | undefined;
  temperatureMeasurementUnit: app.MeasurementUnit;
  isLoading: boolean;
}): JSX.Element => {
  const { t } = useTranslation();
  const { weather, temperatureMeasurementUnit, isLoading } = props;

  return (
    <Box
      margin={{ base: '60px auto', sm: '10px auto', lg: '60px auto' }}
      display="grid"
      gridTemplateColumns={{
        base: '1fr 1fr 1fr',
        sm: '1fr',
        lg: '1fr 1fr 1fr',
      }}
      width="100%"
      alignItems="center"
    >
      <Box
        fontSize={{ base: '8xl', sm: '5xl', md: '7xl', lg: '8xl' }}
        color="white"
        fontFamily="Open Sans"
        lineHeight="90px"
        textAlign="center"
        textShadow="1px 1px 5px rgba(0,0,0,0.3)"
      >
        <SkeletonText isLoaded={!isLoading} noOfLines={1} skeletonHeight="90px">
          <Text>
            {weather?.temperature
              ? getFullTemperature(
                  weather.temperature,
                  temperatureMeasurementUnit
                )
              : ''}
          </Text>
        </SkeletonText>
      </Box>
      <Box textAlign="center" margin="0 auto">
        <Skeleton isLoaded={!isLoading} height="70px" minWidth="70px">
          {weather && (
            <AsyncSvgIcon
              svg={`weather-${weather.weather.toLocaleLowerCase()}`}
              options={{ fill: 'white', width: '70px', height: '70px' }}
            />
          )}
        </Skeleton>
        <SkeletonText
          isLoaded={!isLoading}
          mt={2}
          noOfLines={1}
          spacing="4"
          skeletonHeight="41px"
        >
          <Text
            color="white"
            fontSize="sm"
            fontWeight="bold"
            fontFamily="Montserrat"
            padding="10px 0"
          >
            {weather?.weather ? t(weather.weather) : ''}
          </Text>
        </SkeletonText>
      </Box>
      <Box color="white" textAlign={{ base: 'left', sm: 'center', lg: 'left' }}>
        <Box textShadow="1px 1px 5px rgba(0,0,0,0.3)">
          <SkeletonText
            isLoaded={!isLoading}
            mt="2"
            noOfLines={1}
            spacing="4"
            skeletonHeight="18px"
          >
            {weather?.wind_speed && (
              <>
                <Text fontWeight="bold" as="span">
                  {t('Wind speed: ')}
                </Text>
                {` No ${weather?.wind_speed && weather.wind_speed}km/h`}
              </>
            )}
          </SkeletonText>
        </Box>
        <Box textShadow="1px 1px 5px rgba(0,0,0,0.3)">
          <SkeletonText
            isLoaded={!isLoading}
            mt="2"
            noOfLines={1}
            spacing="4"
            skeletonHeight="18px"
          >
            {weather?.humidity && (
              <>
                <Text fontWeight="bold" as="span">
                  {t('Humidity: ')}
                </Text>
                {` ${weather.humidity}%`}
              </>
            )}
          </SkeletonText>
        </Box>

        <Box textShadow="1px 1px 5px rgba(0,0,0,0.3)">
          <SkeletonText
            isLoaded={!isLoading}
            mt="2"
            noOfLines={1}
            spacing="4"
            skeletonHeight="18px"
          >
            {weather?.pressure && (
              <>
                <Text fontWeight="bold" as="span">
                  {t('Pressure: ')}
                </Text>
                {` ${weather.pressure}hPA`}
              </>
            )}
          </SkeletonText>
        </Box>
      </Box>
    </Box>
  );
};

export default WeatherTodayDetails;
