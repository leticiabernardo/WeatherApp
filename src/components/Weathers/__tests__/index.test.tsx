/* eslint-disable @typescript-eslint/ban-ts-comment */
import { render, cleanup, waitFor } from '@testing-library/react';
import Weathers from '../index';

const current: app.CurrentWeather = {
  date: 1645452000,
  pressure: 25,
  humidity: 7,
  wind_speed: 80,
  weather: 'Clear' as app.WeatherName,
  temperature: 27,
  fullTemperature: '27 °C',
  location: 'Ceará',
};

const daily: app.DailyWeathers[] = [
  {
    date: 1645452000,
    pressure: 25,
    humidity: 7,
    wind_speed: 80,
    weather: 'Clear',
    temp: {
      min: 23,
      max: 32,
    },
  },
];

const temperatureMeasurementUnit: app.MeasurementUnit = 'celsius';
const weathers: app.WeatherResponse = { current, daily };

class MockDate extends Date {
  constructor() {
    super('2020-05-14T11:47:10Z');
  }
}

describe('weathers render', () => {
  afterEach(cleanup);

  test('render the component', async () => {
    // @ts-ignore
    global.Date = MockDate;
    const { container } = await waitFor(() =>
      render(
        <Weathers
          weathers={weathers}
          temperatureMeasurementUnit={temperatureMeasurementUnit}
        />
      )
    );
    expect(container).toMatchSnapshot();
  });
});
