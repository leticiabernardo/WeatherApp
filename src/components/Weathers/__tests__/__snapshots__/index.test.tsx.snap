// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`weathers render render the component 1`] = `
<div>
  <div
    class="css-57i5ee"
  >
    <span
      class="css-0"
    >
      <div>
        <div
          class="css-8atqhb"
        >
          <div
            class="css-1hoatnx"
          >
            <p
              class="chakra-text css-1vza8hn"
            >
              Ceará
            </p>
            <p
              class="chakra-text css-i2jvxk"
            >
              11:47
            </p>
            <p
              class="chakra-text css-xj20wo"
            >
              Today {{date}}
            </p>
          </div>
          <div
            class="css-1uxcmd3"
          >
            <div
              class="css-9nbp22"
            >
              <div
                class="css-y7eb69"
              >
                27 °C
              </div>
              <div
                class="css-rt6i1c"
              >
                <div
                  class="svg-mock"
                />
                <p
                  class="chakra-text css-187qkfd"
                >
                  Clear
                </p>
              </div>
              <div
                class="css-1mrlgg5"
              >
                <p
                  class="chakra-text css-1ac62sw"
                >
                  <span
                    class="chakra-text css-in3yi3"
                  >
                    Wind speed: 
                  </span>
                   No 80km/h
                </p>
                <p
                  class="chakra-text css-1ac62sw"
                >
                  <span
                    class="chakra-text css-in3yi3"
                  >
                    Humidity: 
                  </span>
                   7%
                </p>
                <p
                  class="chakra-text css-1ac62sw"
                >
                  <span
                    class="chakra-text css-in3yi3"
                  >
                    Pressure: 
                  </span>
                   25hPA
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-1r78p7n"
      >
        <div
          class="css-u8prhs"
        >
          <p
            class="chakra-text css-4g3ufm"
          >
            segunda
          </p>
          <div
            class="css-1a82xio"
          >
            <div
              class="svg-mock"
            />
            <p
              class="chakra-text css-1fub2xe"
            >
              Clear
            </p>
          </div>
          <div
            class="css-kwfulx"
          >
            <p
              class="chakra-text css-0"
            >
              Max: {{max}}
            </p>
            <p
              class="chakra-text css-0"
            >
              Min: {{min}}
            </p>
          </div>
        </div>
      </div>
    </span>
  </div>
</div>
`;
